core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1531,56 +1787,75 @@ class ReplicaManager(val config: KafkaConfig,
     val builder = imageBuilder.partitionsBuilder()
     val startMs = time.milliseconds()
     replicaStateChangeLock synchronized {
+      val deferrable = changesDeferrable
       stateChangeLogger.info(("Metadata batch %d: %d local partition(s) changed, %d " +
         "local partition(s) removed.").format(metadataOffset, builder.localChanged().size,
           builder.localRemoved().size))
       if (stateChangeLogger.isTraceEnabled) {
         builder.localChanged().foreach { state =>
-          stateChangeLogger.trace(s"Metadata batch ${metadataOffset}: locally changed: ${state}")
+          stateChangeLogger.trace(s"Metadata batch $metadataOffset: locally changed: ${state}")
         }
         builder.localRemoved().foreach { state =>
-          stateChangeLogger.trace(s"Metadata batch ${metadataOffset}: locally removed: ${state}")
+          stateChangeLogger.trace(s"Metadata batch $metadataOffset: locally removed: ${state}")
         }
       }
       // First create the partition if it doesn't exist already
+      // partitionChangesToBeDeferred maps each partition to be deferred to its (current state, previous deferred state if any)
+      val partitionChangesToBeDeferred = mutable.HashMap[Partition, (MetadataPartition, Option[HostedPartition.Deferred])]()
```

Comment:
The way we do this is a little curious. If `changesDeferrable` is true, then aren't `partitionsToBeLeader` and `partitionsToBeFollower` unnecessary? It seems like we can simplify this and just jump directly to `makeDeferred` in that case. That would make this code easier to follow since the cases would not all be mixed together.